### PR TITLE
Support Custom Service Start Up order

### PR DESCRIPTION
**Pull request by Sgitario**
By default, the services are initialized in Natural Order of presence. For example, having:

```java
class MyParent {
    @QuarkusApplication // ... or @Container
    static final RestService firstAppInParent = new RestService();
    
    @QuarkusApplication // ... or @Container
    static final RestService secondAppInParent = new RestService();

}

@QuarkusScenario
class MyScenarioIT extends MyParent {
    @QuarkusApplication // ... or @Container
    static final RestService firstAppInChild = new RestService();

    @QuarkusApplication // ... or @Container
    static final RestService secondAppInChild = new RestService();
}
```

Then, the framework will initialize the services at this order: `firstAppInParent`,  `secondAppInParent`, `firstAppInChild` and `secondAppInChild`.

We can change this order by using the `@LookupService` annotation:

```java
class MyParent {
    @LookupService
    static final RestService appInChild;

    @QuarkusApplication // ... or @Container
    static final RestService appInParent = new RestService().withProperty("x", () -> appInChild.getHost());
}

@QuarkusScenario
class MyScenarioIT extends MyParent {
    @QuarkusApplication // ... or @Container
    static final RestService appInChild = new RestService().dependsOn(appInParent);
}
```

Now, the framework will initialize the `appInChild` service first and then the `appInParent` service.

### CODE REVIEW

File: quarkus-test-core/src/main/java/io/quarkus/test/utils/ReflectionUtils.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    public static List<Field> findAllFields(Class<?> clazz) {
+        if (clazz == Object.class) {
+            return Collections.emptyList();
+        }
+
+        List<Field> fields = new ArrayList<>();
+        fields.addAll(findAllFields(clazz.getSuperclass()));
+        fields.addAll(Arrays.asList(clazz.getDeclaredFields()));
```
Comment (jsmrcka) [inline quarkus-test-core/src/main/java/io/quarkus/test/utils/ReflectionUtils.java:50]: `clazz.getDeclaredFields()` does not guarantee an order. So we cannot be sure that the services from within a single class will be initialized in the order of presence.
The `@LookupService` can only be used in a parent-child setup, but does not address situations like this:
```java
@QuarkusScenario
class MyScenarioIT {
    @QuarkusApplication
    static final RestService appOne = new RestService();

    @QuarkusApplication
    static final RestService appTwo = new RestService().withProperty("x", () -> appOne.getHost());
}
```

File: README.md [hunk 2]
```diff
@@ -82,6 +82,52 @@ public class PingPongResourceIT {
...
+    static final RestService appInChild;
+
+    @QuarkusApplication // ... or @Container
+    static final RestService appInParent = new RestService().withProperty("x", () -> appInChild.getHost());
+}
+
+@QuarkusScenario
+class MyScenarioIT extends MyParent {
+    @QuarkusApplication // ... or @Container
+    static final RestService appInChild = new RestService().dependsOn(appInParent);
```
[the remark this answers is not in the document]
Comment (Sgitario) [reply README.md:125]: Ops, this is a left over. Thanks for spotting!

File: README.md [hunk 3]
```diff
@@ -82,6 +82,52 @@ public class PingPongResourceIT {
...
+}
+```
+
+Now, the framework will initialize the `appInChild` service first and then the `appInParent` service.
+
```
[the remark this answers is not in the document]
Comment (Sgitario) [reply README.md:130]: +100

File: quarkus-test-core/src/main/java/io/quarkus/test/utils/ReflectionUtils.java (hunk 1 above)
Comment (Sgitario) [reply quarkus-test-core/src/main/java/io/quarkus/test/utils/ReflectionUtils.java:50]: This scenario is already supported. Services are init by order of presence within the same class.

About the `clazz.getDeclaredFields()`, the order is not guaranteed when all the fields are independent from each other. For example:

```
@QuarkusScenario
class MyScenarioIT {
    @QuarkusApplication
    static final RestService appOne = new RestService();

    @QuarkusApplication
    static final RestService appTwo = new RestService();
}
```

Here, it's true that the bytecode output is not guaranteing the field order, but here we don't really care as services do not depend on each other.

For this scenario:

```
@QuarkusScenario
class MyScenarioIT {
    @QuarkusApplication
    static final RestService appOne = new RestService();

    @QuarkusApplication
    static final RestService appTwo = new RestService().withProperty("x", () -> appOne.getHost());
}
```

The field order is guaranteed by the bytecode output (which is what the clazz.getDeclaredFields method is based on).

Comment (jsmrcka) [reply quarkus-test-core/src/main/java/io/quarkus/test/utils/ReflectionUtils.java:50]: You're right, I didn't realize that. Thanks for explaining. :)

**Review by jsmrcka — APPROVED** [review]

**Pull request merged**